core/trino-main/src/main/java/io/trino/likematcher/LikeMatcher.java

```diff
@@ -0,0 +1,368 @@
+            else if (character == '%' || character == '_') {
+                if (literal.length() != 0) {
+                    result.add(new Literal(literal.toString()));
+                    literal = new StringBuilder();
+                }
+
+                if (character == '%') {
+                    result.add(new Any(0, true));
+                }
+                else {
+                    result.add(new Any(1, false));
+                }
+            }
+            else {
+                literal.append(character);
+            }
+        }
+
+        if (inEscape) {
+            throw new IllegalArgumentException("Escape character must be followed by '%', '_' or the escape character itself");
+        }
+
+        if (literal.length() != 0) {
+            result.add(new Literal(literal.toString()));
+        }
+
+        return result;
+    }
+
+    private static List<Pattern> optimize(List<Pattern> pattern)
+    {
+        if (pattern.isEmpty()) {
+            return pattern;
+        }
+
+        List<Pattern> result = new ArrayList<>();
+
+        int anyPatternStart = -1;
+        for (int i = 0; i < pattern.size(); i++) {
+            Pattern current = pattern.get(i);
+
+            if (anyPatternStart == -1 && current instanceof Any) {
+                anyPatternStart = i;
+            }
+            else if (current instanceof Literal) {
+                if (anyPatternStart != -1) {
+                    result.add(collapse(pattern, anyPatternStart, i));
+                }
+
+                result.add(current);
+                anyPatternStart = -1;
+            }
+        }
+
+        if (anyPatternStart != -1) {
+            result.add(collapse(pattern, anyPatternStart, pattern.size()));
+        }
+
+        return result;
+    }
+
+    /**
+     * Collapses a sequence of consecutive Any items
+     */
+    private static Any collapse(List<Pattern> pattern, int start, int end)
+    {
+        int min = 0;
+        boolean unbounded = false;
+
+        for (int i = start; i < end; i++) {
+            Any any = (Any) pattern.get(i);
+
+            min += any.min();
+            unbounded = unbounded || any.unbounded();
+        }
+
+        return new Any(min, unbounded);
+    }
+
+    private static NFA makeNfa(List<Pattern> pattern)
+    {
+        NFA.Builder builder = new NFA.Builder();
+
+        NFA.State state = builder.addStartState();
+
+        for (Pattern item : pattern) {
+            if (item instanceof Literal literal) {
+                for (byte current : literal.value().getBytes(UTF_8)) {
+                    state = matchByte(builder, state, current);
+                }
+            }
+            else if (item instanceof Any any) {
+                NFA.State previous;
+                int i = 0;
+                do {
+                    previous = state;
+                    state = matchSingleUtf8(builder, state);
+                    i++;
+                }
+                while (i < any.min());
+
```

Comment:
nit: I think this works because `any(0, false)` is not allowed by definition because it's just an empty string and not a transition. would leave it to you if worth  enforcing that in Any constructor

Reply:
Yes, that's a good point. I added a check to the Any constructor and adjusted code elsewhere that could result in `any(0, false)` after pattern optimization.